File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.java
class Solution {
    
    public int subarraysWithKDistinct(int[] nums, int k) {
        return dfs(nums,k) - dfs(nums,k-1);
    }

    public int dfs(int[] nums,int k){
        int i=0;
        int ans = 0;
        HashMap<Integer,Integer> d = new HashMap<>();
        for(int j=0;j<nums.length;j++){
            if (!d.containsKey(nums[j])){
                d.put(nums[j],0);
            }
            d.put(nums[j],d.get(nums[j])+1);
            while(d.size()>k){
                d.put(nums[i],d.get(nums[i]) -1);
                if(d.get(nums[i]) == 0){
                    d.remove(nums[i]);
                }
                i+=1;
            }
            ans += j-i+1;
        }
        return ans;
    }
}

File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.py
class Solution:
    def subarraysWithKDistinct(self, s: List[int], k: int) -> int:
       

        def dfs(t):
            d ={}
            i= 0 
            ans = 0
        
            for j in range(len(s)):
                d[s[j]] = d.get(s[j],0) +1
                while len(d)> t:
                    d[s[i]]-=1
                    if d[s[i]] == 0:
                        d.pop(s[i])
                    i+=1
                ans += j-i+1
            return ans
        return dfs(k) - dfs(k-1)

        